ErrorableRadioButtons component (HCA Waffle #97)

joesoftware opened the pull request:
Implementation of an ErrorableRadioButtons component, with one instance of it used. Tests updated and validations updated for the question that is now using radio buttons. Styles may need to be adjusted in the future if there is ever a use of a group of radio buttons that should not be displayed inline.

joesoftware committed:
Add radio button component for child info.

Change child relationship back to a select component.
Change has child to report from checkbox to radio button.
Fix radio button component to properly use React's id and htmlFor properties.
Clarify usage of value in radio button component.
Add yes/no array to select (and now radio button) constants.
Alphabetize child relationship constants.

Fixed component to allow booleans and strings to match value in app datastore.
Fixed most of the spec changes (and removed the irrelevant ones). `xit(` the ones that require more research.
Removed the required and error state.

Add missing ErrorableCheckbox.

Rename the component. Simplify checked-button logic.

Added renamed files to repo.

Make button 'require-able'.

akainic commented on _health-care/_js/components/financial-assessment/Child.jsx:
```diff
@@ -29,7 +29,7 @@ class Child extends React.Component {
         <div className="row">
           <div className="small-12 columns">
-            <ErrorableSelect required
+            <ErrorableRadioButtonGroup required
```
I don't think this is the right question we want to make into a radio button. It's the checkbox question "Do you have any children to report?" in `ChildInformationSection.jsx`.

joesoftware replied:
Cool. I peeked in our pre-React application and picked the first radio button example I saw which was relationship. I'll move this back into the parent component.

akainic commented on _health-care/_js/components/form-elements/ErrorableRadioButtonGroup.jsx:
```diff
@@ -0,0 +1,83 @@
+      return (
+        <div key={reactKey++}>
+          <input
+              type="radio"
```
The input needs a unique `id` and the label needs a `for` attribute that matches the `id` of the input. This will associate the label with the input so that if someone clicks the label the input will toggle.

akainic commented on _health-care/_js/components/form-elements/ErrorableRadioButtonGroup.jsx:
```diff
@@ -0,0 +1,83 @@
+        label = obj.label;
+        value = obj.value;
+      }
+      const checked = value === this.props.value ? 'checked=true' : '';
```
I think we won't have access to `this.props.value` once inside the `map` function. Set this value to a variable outside of `map` and then access it through the variable. Maybe we should have more descriptive variable names for the 2 different values we have: `storedValue` for `this.props.value` and `inputValue` for `value`?

joesoftware replied:
Sounds good.

akainic commented on _health-care/_js/components/form-elements/ErrorableRadioButtonGroup.jsx:
```diff
@@ -0,0 +1,83 @@
+              checked={checked}
+              name={this.inputId}
+              value={value}
+              onChange={this.handleChange}/>
```
I think the reason this isn't selectable is because the `onChange` isn't getting triggered. Not sure why that is yet.

akainic commented on _health-care/_js/components/financial-assessment/ChildInformationSection.jsx:
```diff
@@ -49,7 +51,10 @@ class ChildInformationSection extends React.Component {
+    // TODO: Yes/No is a weird case for the related radio buttons, this should probably be boolean.
+    //        Should we make a specific case for this?
+    if (this.props.data.hasChildrenToReport === true ||
+        this.props.data.hasChildrenToReport === 'Yes') {
```
We should only have one possible stored value, not two. The reason why we were checking for `true` before was because it was a checkbox. We could have the labels for the radio buttons be `Yes` and `No` and still store the value as `true` and `false`. Let's follow whatever the original form does in terms of the value and stick with that.

akainic commented:
The radio buttons need a bit of styling help. Do you mind if I push a commit to this branch to add the right styles?

akainic commented:
Actually nevermind I'll fix the styles in a follow-up PR :)

akainic commented on spec/javascripts/health-care/components/form-elements/ErrorableRadioButtonGroup.spec.jsx:
```diff
@@ -0,0 +1,166 @@
+      sinon.assert.calledWithMatch(consoleStub, /Invalid prop `onValueChange` of type `boolean` supplied to `ErrorableRadioButtonGroup`, expected `function`/);
+    });
+
+    /*
```
Can we put an `x` in front of `it` on line 43, as well as for other tests, in order to skip them instead of commenting them out?

joesoftware replied:
Sensible. I'd forgotten about `xit`.

akainic commented on spec/javascripts/health-care/components/form-elements/ErrorableRadioButtonGroup.spec.jsx:
```diff
@@ -0,0 +1,166 @@
+      );
+    });
+
+    const input = ReactTestUtils.findRenderedDOMComponentWithTag(errorableRadioButtonGroup, 'input');
```
We're going to have to change this test from checking for any input and instead checking for the one that's checked.

joesoftware replied:
Copy pasta error.

akainic commented on _health-care/_js/components/financial-assessment/ChildInformationSection.jsx:
```diff
@@ -66,9 +71,10 @@ class ChildInformationSection extends React.Component {
 
         <div className="input-section">
-          <ErrorableCheckbox
+          <ErrorableRadioButtonGroup
```
Now that this is a radio button, should this field be required? Can we assume if a user does not select either "yes" or "no" that they do not have children to report?

joesoftware replied:
Agreed.

akainic replied:
Right, I think it wasn't required before when it was a checkbox because not checking a checkbox implies "No", but with a radio button you have to select something to answer the question.

akainic commented on spec/javascripts/health-care/components/form-elements/ErrorableRadioButtonGroup.spec.jsx:
```diff
@@ -0,0 +1,166 @@
+    expect(labels[0].text()).to.equal('my label');
+
+    // Ensure label htmlFor is attached to select id.
+    const inputs = tree.everySubTree('select');
```
This test needs to be updated to not check for `select`s.

akainic commented:
I know some of the tests aren't passing now, which is why they are commented out, but we should try to figure out how to get them to pass before merging this.

akainic commented on spec/javascripts/health-care/components/form-elements/ErrorableRadioButtonGroup.spec.jsx:
```diff
@@ -0,0 +1,86 @@
+    });
+  });
+
+  xit('ensure value changes propagate', () => {
```
Let's try to get these tests that are commented out passing before we merge this.

akainic commented on _health-care/_js/components/form-elements/ErrorableRadioButtonGroup.jsx:
```diff
@@ -0,0 +1,98 @@
+        optionValue = obj.value;
+      }
+      let checked;
+      checked = (storedValue && optionValue.toString() === storedValue.toString()) ? checked = 'checked=true' : '';
```
This is a little wonky. Why do you have to convert both of these to strings? They should both be stored as booleans, so you should be able to just compare them. I think all you need to replace lines 51-52 is:
```javascript
const checked = storedValue !== undefined && optionValue === storedValue ? 'checked=true' : '';
```

I'm assuming you have the initial `storedValue` check to make sure it's not undefined? Unfortunately we have to actually check if it's not undefined because simply calling it will return it's value, which, if set, is either true or false, which will affect what this whole expression evaluates to.

joesoftware replied:
Good catch.

joesoftware commented:
The radio button usage has been elided from the current code. This remains as an example for later use.

akainic committed:
Fix tests, add one instance of ErrorableRadioButton, and add correct styling

Implement one instance of ErrorableRadioButtons

Style radio buttons, and make sure error messages actually appear

joesoftware commented on _health-care/_js/components/form-elements/ErrorableRadioButtons.jsx:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  handleChange(domEvent) {
+    this.props.onValueChange(domEvent.target.value);
```
If we allow boolean values, then this handler will need to make sure the `<input>` element doesn't enthusiastically change the value back to a string.

akainic replied:
Agreed, let's deal with that if it turns out that we in fact need to store boolean values

joesoftware commented on _health-care/_js/components/form-elements/ErrorableRadioButtons.jsx:
```diff
@@ -0,0 +1,112 @@
+        label: React.PropTypes.string,
+        value: React.PropTypes.oneOfType([
+          React.PropTypes.string,
+          React.PropTypes.bool
```
Long term, do we want to allow non-string types for this component? The `<input type="radio">` takes liberties converting things to strings.

joesoftware commented:
LGTM with comments for later design discussion.
